README.md

```diff
@@ -1,2 +1,2 @@
-# resource-dispatcher-image
-Image for Kubernetes meta controller server which generates Kubernetes resources which need to be injected to user namespaces. 
+# Resource dispatcher Docker image
+This is an image for resource dispatcher.
```

Comment:
Can we expand this to describe what the tool does?

Reply:
This is a great thing to break into a separate PR as well.  It can be standalone

Reply:
I agree this is a relatively small PR functionality wise, but that's the point.  There's 29 files changed here, so it looks and reviews like a big PR even though it could be a few small PRs.

imo (and I bet if she was here, in @DnPlas's opinion) we can do much better at splitting PRs and making these reviews simpler.  Adding a README.md as a separate PR is really easy, so I'd prefer it be done that way.  But that isn't a showstopper here.

Making separate PRs sounds harder, and it is harder *after the fact*.  But if done as the work progresses it is easier on everyone.  The boilerplate (copied tox.ini, stuff like that) could be a PR that's reviewed right away.  The actual resource dispatcher script and its tests can be their own PR.  The Dockerfile/CI that builds the image can be a separate PR. These all have distinct functionality that can be reviewed individually

Reply:
I agree with @ca-scribner. There are a couple of changes that definitely belong in other PRs. I can understand that splitting after submitting a large number of changes can be hard, and it's hard to be fair.

To me changes in this PR could be grouped in:
* Group 1: boilerplate (tox.ini, requirements*, workflows, .gitignore, etc.)
* Group 2: documentation changes
* Group 3: src and tests, and I'd argue these could even be in separate PRs

As @ca-scribner said, this does not have to be a showstopper, but it would be nice.

Reply:
Agree - I don't think this is a huge deal and given that there has been a fair bit of review already, probably just one for next time.

One thing you *could* do is rebase this PR into 3 clear commits (according to Daniela's suggestion above) and rebase rather than squash merge, but again shouldn't be a blocker a don't think.

Reply:
From our best practices:
> verifiable chunks that can function separately is very helpful

Except documentation and workflows, the proposed PR break down will go against exactly that requirement: Group 1 will fail without Group 3. Group 3 will not have way to run tests, because `tox.ini` updates are in Group 1. And Group 1 will not have source code to test.
It is impossible to verify Group 1 without Group 3. May be a review of just changes is clearer, but those two pieces won't function by themselves.
I don't buy argument about test driven development on this subject, because it is not development, but code commit strategy. During development it is perfectly fine to start with `tox.ini` changes to make test fail and start building up functionality from that. Separating code later for commits is counterproductive.
Putting tox/source/tests together gives exactly what we want: a verifiable chunk that can function.
In this particular case of brand new charm submitting separate PRs make even reviewing unproductive, because I would need to review two PRs separately and merge them both to achieve testable commit. So, at the end I will be reviewing two PRs merged together anyway.

I am not against small PR strategy, but in this case it is not really possible to avoid large PR in order for it to be testable.
Merging PRs that we know will fail is a dangerous and we should avoid it,

Reply:
I agree that merging things that we know will fail **into main** is dangerous and we should not do that.  As discussed earlier today, that is not what is being discussed.  

What is being said is that, for things that can clearly be separated into independent concerns (eg: the groups @DnPlas mentioned) we should create separate PRs.  If these PRs do not result in a working final product on their own, **they should be opened against a feature branch**.  This way they can each be discussed on their own merits and merged when they're ready.

@jnsgruk you're right that we can separate these different concerns into different commits.  100% agree.  A reason why imo that is less effective than different PRs against a feature branch is because as soon as reviews start, those three separate commits can be updated, force pushed, etc.  Every time a reviewer comes back to the review, they need to interrogate the PR to understand what changed.  By keeping things in separate PRs, the effort on the reviewer is reduced and we can merge non-controversial things quickly.

Reply:
My point is simple, common approach (outlined in our best practices) is for PR to be testable (regardless of size). When we commit separate PRs that cannot be tested, I am not sure code review is valid. Reviewing code that is not working is wasting time, unless it is a draft. Because, later on, when everything is merged (all small PRs) new review needs to be done or when feature branch is merged into `main` someone will have to do it: *The sum of the parts does not equal the whole* 
We do not have formal QA that would verify final functionality for us. I doubt anyone goes and tests end-to-end functionality of final build of `main` branch.

Reply:
@i-chvets 
going into the PR files I see the common base that we use (same as in other charms)
let's take concrete example:
1. renovate -> can easily go into another PR
2. readme, the same
3. tox.ini, the same
4. tests requirements , the same
5. dockerfile, the same
6. gitignore, the same
7. pyproject.toml, the same

these 7 items could be addressed before even writing the code.
and then you have a single PR with all the tests and python code, that is self contained and testable.

Reply:
1. Agreed.
2. Agreed (partially). However, to understand what I am reviewing in other small PRs I would need to look into README, thus switching PRs... deafting purpose of small PRs.
3. Disagree: Executing `tox` commands will fail without source code and tests. What am I reviewing then? It is nicely formatted `tox.ini` file, but does it work? I have no idea. Show me how I can test that `tox.ini` without source and tests, and I will switch disagree to agree. If this `tox.ini` is a boilerplate, let's automate it, so we don't have to review it. For now to test it, I need to bring in "source/test only" PR to verify that `tox.ini` is functional. Thus defeating the purpose of small PRs.
4. Disagree. requirements.* go with the code. If I need to bring some package that I need in my source code, I will have to commit requirements change along with the code, otherwise "source/test only" PR will fail and I won't be able to test it. I would have to bring PR with requirements into the fold, thus defeating the purpose of small PR.
5. Disagree (partially). Dockerfile is useless without the source code. I would have it in a separate PR for it at later stage (after source code is checked in), but at this point I would need to wait for all other PRs to be merged before reviewing Dockerfile.
6. Agree.
7. Agree.